Feature/get-coordinates

am17jx opened the pull request:
… model and mock implementation

- Created Coordinates data model
- Added GetCoordinatesFromCityUseCaseRepo interface in logic.repository
- Implemented GetCoordinatesFromCityRepositoryImpl in data.repository
- Added use case class GetCoordinatesFromCityUseCase in logic.useCase

am17jx committed:
feat: add GetCoordinatesFromCity use case with repository, interface, model and mock implementation

- Created Coordinates data model
- Added GetCoordinatesFromCityUseCaseRepo interface in logic.repository
- Implemented GetCoordinatesFromCityRepositoryImpl in data.repository
- Added use case class GetCoordinatesFromCityUseCase in logic.useCase

saraashraf11 commented on src/main/kotlin/logic/useCase/GetCoordinatesFromCityUseCase.kt:
```diff
@@ -0,0 +1,12 @@
+    private val repository: GetCoordinatesFromCityRepository
+) {
+    fun execute(city: String): Coordinates? {
+        return repository.getCoordinatesForCity(city)
```
use invoke function as we agreed

saraashraf11 commented on src/main/kotlin/data/repository/GetCoordinatesFromCityRepositoryImpl.kt:
```diff
@@ -0,0 +1,15 @@
+    override fun getCoordinatesForCity(city: String): Coordinates? {
+        return when (city.lowercase()) {
+            //fake data
+            "cairo" -> Coordinates(30.0444, 31.2357)
```
delete this comment

FaresM0hamed replied:
I don't know exactly what is business requirements need here , but i think this approach is not useful , these are only two countries , what if user write city not found in this tiny list  , what will happen ? 
In my opinion you need to find other outbox solution for this implementation , what do you think ?

am17jx replied:
Thanks for the feedback
You’re totally right — this mock is temporary just to unblock the rest of the flow.
I’m currently working on integrating the real geocoding API to handle any city input dynamically.
Once that’s done, this hardcoded list will be replaced. I’ll keep you updated once the API part is ready.

saraashraf11 commented:
I think the name branch needs to be renamed according to your feature, what do you think?

am17jx committed:
Edit : delete commint

Pixelh97 reviewed (commented):
Before i review i need to know the flow for this application,
Also, where is the test cases for this repo and usecase?

MuhammedEdrees commented on src/main/kotlin/data/repository/GetCoordinatesFromCityRepositoryImpl.kt:
```diff
@@ -0,0 +1,15 @@
+import org.example.logic.repository.GetCoordinatesFromCityRepository
+import org.example.logic.model.Coordinates
+
+class GetCoordinatesFromCityRepositoryImpl : GetCoordinatesFromCityRepository {
```
I think naming convention here is bad. Classes normally shouldn't start with a verb and having the repo name too similar to its function name is too specific

am17jx committed:
refactor: rename GetCoordinatesFromCityRepository to CoordinatesFromCityRepository interface

am17jx committed:
feat: implement and rename CoordinatesFromCityRepositoryImpl

- Renamed from GetCoordinatesFromCityRepositoryImpl to CoordinatesFromCityRepositoryImpl
- Added implementation using Open-Meteo Geocoding API
- Fetches coordinates (latitude, longitude) based on city name

am17jx committed:
Merge branch 'develop' into feature/get-coordinates

am17jx committed:
feat: implement CoordinatesFromCityRepositoryImpl

MuhammedEdrees commented on src/main/kotlin/data/repository/CoordinatesFromCityRepositoryImpl.kt:
```diff
@@ -0,0 +1,32 @@
+    override fun getCoordinatesForCity(city: String): Coordinates? {
+        val url = "https://geocoding-api.open-meteo.com/v1/search?name=${city}&count=1"
+
+        val client = OkHttpClient()
+        val request = Request.Builder()
+            .url(url)
+            .build()
```
having the client created inside the function makes it hard to test it should be extracted as constructor property and be injected using DI.

MuhammedEdrees reviewed (changes requested):
I noticed that you are using OkHttp for http requests but in the same repo and on develop branch Ktor is already in use. There should be consistency in the libraries used for the same purpose across the whole project.

am17jx committed:
refactor: rewrite CoordinatesFromCityRepositoryImpl using ktor

FaresM0hamed commented on src/main/kotlin/data/repository/CoordinatesFromCityRepositoryImpl.kt:
```diff
@@ -0,0 +1,34 @@
+) : CoordinatesFromCityRepository {
+
+    override fun getCoordinatesForCity(city: String): Coordinates? = runBlocking {
+        val url = "https://geocoding-api.open-meteo.com/v1/search?name=$city&count=1"
```
I prefer to move this to constants , what do you think ?

am17jx replied:
Thank you for the suggestion Since the project is still small and scoped, I preferred to keep the variables inline for now to keep things simple and focused.

Pixelh97 commented on src/main/kotlin/logic/repository/CoordinatesFromCityRepository.kt:
```diff
@@ -0,0 +1,6 @@
+import org.example.logic.model.Coordinates
+
+interface CoordinatesFromCityRepository {
+    fun getCoordinatesForCity(city: String): Coordinates?
```
Since we are using coroutines, you should make this function suspend

FaresM0hamed commented on src/main/kotlin/logic/useCase/GetCoordinatesFromCityUseCase.kt:
```diff
@@ -0,0 +1,12 @@
+class GetCoordinatesFromCityUseCase(
+    private val repository: CoordinatesFromCityRepository
+) {
+    fun invoke(city: String): Coordinates? {
```
This fun should be suspend

Pixelh97 replied:
plus, this should have the `operator` modifier to run the `invoke` function

Pixelh97 commented on src/main/kotlin/data/repository/CoordinatesFromCityRepositoryImpl.kt:
```diff
@@ -0,0 +1,34 @@
+
+        val locationObject = results.getJSONObject(0)
+        val lat = locationObject.optDouble("latitude", Double.NaN)
+        val lon = locationObject.optDouble("longitude", Double.NaN)
```
avoid using a magic numbers and hard coded string,
also why are you using this json structure?

Pixelh97 commented on src/main/kotlin/data/repository/CoordinatesFromCityRepositoryImpl.kt:
```diff
@@ -0,0 +1,34 @@
+    private val client: HttpClient
+) : CoordinatesFromCityRepository {
+
+    override fun getCoordinatesForCity(city: String): Coordinates? = runBlocking {
```
And you shouldn't use the `runBlocking` here you have to make it suspend

Pixelh97 reviewed (changes requested):
One thing to be mentioned, you need to rename your branch to be more detailed

am17jx committed:
Merge branch 'develop' into feature/get-coordinates

am17jx committed:
refactor: update invoke to be suspend and operator

MuhammedEdrees commented on src/main/kotlin/data/repository/CoordinatesFromCityRepositoryImpl.kt:
```diff
@@ -0,0 +1,37 @@
+import org.example.logic.model.Coordinates
+import org.example.logic.repository.CoordinatesFromCityRepository
+
+@Serializable
+data class GeocodingResponse(
+    val results: List<GeoResult>?
+)
+
+@Serializable
+data class GeoResult(
+    val latitude: Double,
+    val longitude: Double
+)
```
Move into separate files

am17jx committed:
refactor: split models into separate files

Pixelh97 reviewed (approved):
Great job

MuhammedEdrees reviewed (approved):
Great effort, well done 👏
